src/backend/ManagementHub.Service/Areas/Tournaments/TournamentsController.cs

```diff
@@ -688,6 +761,281 @@ await this.tournamentContextProvider.RemoveTeamParticipantAsync(
+			}).ToList(),
+			Coaches = model.Coaches.Select(c => new RosterStaffData
+			{
+				UserId = UserIdentifier.Parse(c.UserId)
+			}).ToList(),
+			Staff = model.Staff.Select(s => new RosterStaffData
+			{
+				UserId = UserIdentifier.Parse(s.UserId)
+			}).ToList()
+		};
+
+		// Validate jersey numbers are unique
+		var duplicateNumbers = rosterData.Players
+			.GroupBy(p => p.JerseyNumber)
+			.Where(g => g.Count() > 1)
+			.Select(g => g.Key)
+			.ToList();
+
+		if (duplicateNumbers.Any())
+		{
+			return this.BadRequest(new
+			{
+				error = $"Duplicate jersey numbers: {string.Join(", ", duplicateNumbers)}"
+			});
+		}
+
+		// Update roster
+		try
+		{
+			await this.tournamentContextProvider.UpdateParticipantRosterAsync(
+				tournamentId, teamId, rosterData, this.HttpContext.RequestAborted);
+		}
+		catch (NotFoundException)
+		{
+			return this.NotFound(new { error = "Team is not a participant" });
+		}
+		catch (InvalidOperationException ex)
+		{
+			return this.BadRequest(new { error = ex.Message });
+		}
+
+		return this.Ok();
+	}
+
+	// Helper methods
+
+	private async Task<Dictionary<UserIdentifier, string?>> GetAccessibleGenderDataAsync(
+		List<UserIdentifier> userIds,
+		IUserContext currentUser)
+	{
+		// First, batch check which users the current user can access
+		var accessibleUserIds = new List<UserIdentifier>();
+
+		// Add current user's own ID if it's in the list
+		if (userIds.Contains(currentUser.UserId))
+		{
+			accessibleUserIds.Add(currentUser.UserId);
+		}
+
+		// Get database IDs for all users at once
+		var userIdStrings = userIds.Select(id => id.ToString()).ToList();
+		var userIdLegacyIds = userIds.Select(id => id.ToLegacyUserId()).ToList();
+
+		var userMapping = await this.dbContext.Users
+			.Where(u => (u.UniqueId != null && userIdStrings.Contains(u.UniqueId)) || (u.UniqueId == null && userIdLegacyIds.Contains(u.Id)))
+			.Select(u => new
+			{
+				DbId = u.Id,
+				Identifier = u.UniqueId != null ? UserIdentifier.Parse(u.UniqueId) : UserIdentifier.FromLegacyUserId(u.Id)
+			})
+			.ToListAsync(this.HttpContext.RequestAborted);
+
+		var userDbIds = userMapping.Select(m => m.DbId).ToList();
+		var userDbIdToIdentifier = userMapping.ToDictionary(m => m.DbId, m => m.Identifier);
+
+		// Check if current user is tournament manager
+		var tournamentManagerRole = currentUser.Roles.OfType<TournamentManagerRole>().FirstOrDefault();
+		if (tournamentManagerRole != null)
+		{
+			// Batch query: find all players in tournaments managed by current user
+			var playersInManagedTournaments = await this.dbContext.TournamentTeamRosterEntries
+				.Where(entry => userDbIds.Contains(entry.UserId) && entry.Role == RosterRole.Player)
+				.Join(
+					this.dbContext.TournamentTeamParticipants,
+					entry => entry.TournamentTeamParticipantId,
+					participant => participant.Id,
+					(entry, participant) => new { entry.UserId, participant.TournamentId })
+				.Join(
+					this.dbContext.TournamentManagers.Where(tm =>
```

Comment:
Instead of doing custom Where queries on user id you should prefer to join against Users.WithIdentifier(userId) so this logic is kept to one place.